### PR TITLE
Mccalluc/pydoc

### PR DESCRIPTION
**Pull request by mccalluc**
@SFD5311 : Thanks for being interested in improving the documentation: I'm really happy to have this, I just want to make sure we're doing in the usual way, instead of inventing some new mechanism users would need to know. There are prettier pydoc renderers, but this is a start, and shows what we get for free.

I think this could be merged as-is, or the TODOs could be filled in first. Your call.

### CODE REVIEW

**Comment by SFD5311** [comment]
@mccalluc I'll take a stab at filling in those TODOs.

**Comment by SFD5311** [comment]
@mccalluc Filled in some method docstrings here, including kwarg descriptions specific to each method type.  Heavy focus so far on which parameters to include and what the constraints on parameters are.  Haven't yet thought of a good way to work in query semantics.  Would appreciate your thoughts on what's there, though.

File: hubmap_api_py_client/external.py [hunk 1]
```diff
@@ -163,12 +163,83 @@ def _add_method(output_type, ResultsSetSubclass, doc):
...
-    'cluster': 'TODO: document cluster',
-    'dataset': 'TODO: document dataset',
-    'protein': 'TODO: document protein'
+    'cell':
+    '''
+    Select a set of cells. If no params are provided, selects the set of all cells.
+    Otherwise, selects a set of cells filtered based on parameters supplied.
+        where->str The type of entity being supplied as input to query, must be one of
+            ["cell", "gene", "protein", "organ", "dataset"]
+        has->List[str] A list of entities or expressions supplied as input to the query
```
Comment (mccalluc) [inline hubmap_api_py_client/external.py]:
```suggestion
        has->List[str] A list of UUIDs, or of expressions like "GENE > 0.2".
```

File: hubmap_api_py_client/external.py [hunk 2]
```diff
@@ -163,12 +163,83 @@ def _add_method(output_type, ResultsSetSubclass, doc):
...
-    'organ': 'TODO: document organ',
-    'gene': 'TODO: document gene',
-    'cluster': 'TODO: document cluster',
-    'dataset': 'TODO: document dataset',
-    'protein': 'TODO: document protein'
+    'cell':
+    '''
+    Select a set of cells. If no params are provided, selects the set of all cells.
+    Otherwise, selects a set of cells filtered based on parameters supplied.
+        where->str The type of entity being supplied as input to query, must be one of
```
Comment (mccalluc) [inline hubmap_api_py_client/external.py]: Is this a standard format for documenting arguments? There are several out there that tool chains will support. I'd suggest [Google style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html), unless there's a different standard you would prefer.

File: hubmap_api_py_client/external.py [hunk 3]
```diff
@@ -163,12 +163,83 @@ def _add_method(output_type, ResultsSetSubclass, doc):
...
+            ["cell", "gene", "protein", "organ", "dataset"]
+        has->List[str] A list of entities or expressions supplied as input to the query
+        genomic_modality->str Modality to consider in quantitative queries.
+            Required for queries in which "where" is "gene". Must be one of ["rna", "atac"].
```
Comment (mccalluc) [inline hubmap_api_py_client/external.py]: Are there quantitative queries besides "gene"? ie, is it only used for "gene"?

File: hubmap_api_py_client/external.py [hunk 4]
```diff
@@ -163,12 +163,83 @@ def _add_method(output_type, ResultsSetSubclass, doc):
...
+        logical_operator->str The logical operator applied to filters generated by elements of has.
+            Required for queries in which "where" is "gene" or "protein" and
```
Comment (mccalluc) [inline hubmap_api_py_client/external.py]: We supply a default value, so not really required by Client.

File: hubmap_api_py_client/external.py [hunk 5]
```diff
@@ -163,12 +163,83 @@ def _add_method(output_type, ResultsSetSubclass, doc):
...
+            "has" contains more than one element. Must be one of ["and", "or"].
+    ''',
+    'organ':
+    '''
+    Select a set of organs. If no params are provided, selects the set of all organs.
+    Otherwise, selects a set of organs filtered based on parameters supplied.
+        where->str The type of entity being supplied as input to query. Must be one of
+            ["gene", "organ"].
+        has->List[str] A list of entities supplied as input to the query
```
Comment (mccalluc) [inline hubmap_api_py_client/external.py]: more precise, provide example?

File: hubmap_api_py_client/external.py [hunk 6]
```diff
@@ -163,12 +163,83 @@ def _add_method(output_type, ResultsSetSubclass, doc):
...
+        genomic_modality->str Modality to consider for quantitative queries.
+            Required for queries in which "where" is "gene". Must be one of ["rna", "atac"].
+        p_value->float Threshold of significance applied to gene-organ associations.
+            Required for queries in which "where" is "gene." Must be between 0 and 1.
```
Comment (mccalluc) [inline hubmap_api_py_client/external.py:206]: Beyond being required, is it only used for "gene"?

**Review by mccalluc — COMMENTED** [review]
I think this could be merged as is, but:
- Prefer to use some standard format for this information, for compatibility with tool chains. (if that's already the case, great, I just didn't recognize the style.)
- I think it would be good to spell out what is meant by "entity".

With the list of valid kwargs right there, I can try to make the method signatures match the docs.
Thanks for tackling this!

File: hubmap_api_py_client/external.py (hunk 3 above)
Comment (SFD5311) [reply hubmap_api_py_client/external.py]: For organ, the only quantitative queries are ones in which where is gene.

File: hubmap_api_py_client/external.py (hunk 2 above)
Comment (SFD5311) [reply hubmap_api_py_client/external.py]: I was primarily focused on content for this draft.  I'm happy to move to Google style.

File: hubmap_api_py_client/external.py (hunk 1 above)
Comment (SFD5311) [reply hubmap_api_py_client/external.py]: While the identifiers are universally unique, I feel like the use of UUID, especially in a HuBMAP context has a specific connotation that's not intended here.  Maybe entity identifiers instead?

File: hubmap_api_py_client/external.py (hunk 5 above)
Comment (SFD5311) [reply hubmap_api_py_client/external.py]: Sure, I wasn't sure how much we wanted to get into examples in the doc string, but I can include some.

File: hubmap_api_py_client/external.py (hunk 6 above)
Comment (SFD5311) [reply hubmap_api_py_client/external.py:206]: Only for "gene", yes.

**Comment by SFD5311** [comment]
I definitely think it makes sense to clarify what's meant by entity.  I'll think on how best to do that and work that in when I reformat.

File: hubmap_api_py_client/external.py (hunk 1 above)
Comment (mccalluc) [reply hubmap_api_py_client/external.py]: Sure: or just plain "IDs" would be fine, too. I think of entities as things that are referenced by ID, so I was just a little confused by that part in the original phrasing.

File: hubmap_api_py_client/external.py (hunk 5 above)
Comment (mccalluc) [reply hubmap_api_py_client/external.py]: Thanks: I think this is just another case where "entities" doesn't tell me a lot: I don't think we need to quote actual IDs, but we do need to tell users what information they already have can be used here.

File: hubmap_api_py_client/external.py (hunk 6 above)
Comment (mccalluc) [reply hubmap_api_py_client/external.py:206]: Cool: The if-and-only-if can be more clear.

**Comment by mccalluc** [comment]
Runs on 3.6 are failing because pydoc had a slightly different output format. I'll make a commit to clean that up. Sorry for the failures!

**Comment by mccalluc** [comment]
@SFD5311 : Looks great! I'll going to tweak the links on the README (I forgot that since we use the same MD on pypy, the links need to be absolute) and then merge.

**Review by SFD5311 — APPROVED** [review]

**Pull request merged**